core-blocks/embed/index.js

```diff
@@ -86,43 +82,85 @@ function getEmbedBlockSettings( { title, description, icon, category = 'embed',
 			}
 		},
 
-		edit: class extends Component {
+		edit: compose(
+			withSelect( ( select, ownProps ) => {
+				const { url } = ownProps.attributes;
+				// Preview is undefined if we don't know the status of it, false if it failed,
+				// otherwise it will be an object containing the embed response.
+				const preview = url ? select( 'core' ).getPreview( url ) : undefined;
+				return {
+					preview,
+				};
+			} )
+		)( class extends Component {
 			constructor() {
 				super( ...arguments );
-
-				this.doServerSideRender = this.doServerSideRender.bind( this );
-
+				this.setUrl = this.setUrl.bind( this );
+				this.processPreview = this.processPreview.bind( this );
 				this.state = {
 					html: '',
 					type: '',
 					error: false,
 					fetching: false,
 					providerName: '',
+					url: '',
 				};
 			}
 
 			componentWillMount() {
-				this.doServerSideRender();
+				if ( this.props.attributes.url ) {
+					// Loading from a saved block, set the state up and display the fetching UI.
+					this.setState( { fetching: true, url: this.props.attributes.url } );
+					if ( this.props.preview ) {
+						// Preview must have already been fetched prior to loading this block, so process it.
+						this.processPreview( this.props.preview, this.props.attributes.url );
+					}
+				}
 			}
 
 			componentWillUnmount() {
 				// can't abort the fetch promise, so let it know we will unmount
 				this.unmounting = true;
 			}
 
+			componentWillReceiveProps( nextProps ) {
+				const hasPreview = undefined !== nextProps.preview;
+				if ( hasPreview ) {
+					this.processPreview( nextProps.preview, nextProps.attributes.url );
+				}
+			}
+
 			getPhotoHtml( photo ) {
 				// 100% width for the preview so it fits nicely into the document, some "thumbnails" are
 				// acually the full size photo.
 				const photoPreview = <p><img src={ photo.thumbnail_url } alt={ photo.title } width="100%" /></p>;
 				return renderToString( photoPreview );
 			}
 
-			doServerSideRender( event ) {
+			setUrl( event ) {
 				if ( event ) {
 					event.preventDefault();
 				}
-				const { url } = this.props.attributes;
+				const { url } = this.state;
 				const { setAttributes } = this.props;
+				if ( url === this.props.attributes.url ) {
+					// Don't change anything, otherwise we go into the 'fetching' state but never
+					// get new props, because the url has not changed.
+					return;
+				}
+				setAttributes( { url } );
+				this.setState( { fetching: true, error: false } );
```

Comment:
May be fine to address separately, but see `isRequestingTerms` in `packages/core-data/src/selectors.js` for an example of how we can rely on the resolution flow to determine whether a fetch is happening rather than using state.